pydantic_ai_slim/pydantic_ai/ui/_web/api.py

```diff
@@ -0,0 +1,114 @@
+    builtin_tools: list[BuiltinToolInfo]
+
+
+class _ChatRequestExtra(BaseModel, extra='ignore', alias_generator=to_camel):
+    """Extra data extracted from chat request."""
+
+    model: str | None = None
+    """Model ID selected by the user, e.g. 'openai:gpt-5'. Maps to JSON field 'model'."""
+    builtin_tools: list[str] = []
+    """Tool IDs selected by the user, e.g. ['web_search', 'code_execution']. Maps to JSON field 'builtinTools'."""
+
+
+def add_api_routes(
+    app: Starlette,
+    agent: Agent,
+    models: list[ModelInfo] | None = None,
+    builtin_tools: list[AbstractBuiltinTool] | None = None,
+    toolsets: Sequence[AbstractToolset[Any]] | None = None,
+) -> None:
+    """Add API routes to a Starlette app.
+
+    Args:
+        app: The Starlette app to add routes to.
+        agent: Agent instance.
+        models: Optional list of AI models. If not provided, the UI will have no model options.
+        builtin_tools: Optional list of builtin tools. If not provided, no tools will be available.
+        toolsets: Optional list of toolsets (e.g., MCP servers). These provide additional tools.
+    """
+    _models = models or []
+    _model_ids = {m.id for m in _models}
+    _builtin_tools = builtin_tools or []
+    _toolsets = list(toolsets) if toolsets else None
+    _tools_by_id: dict[str, AbstractBuiltinTool] = {tool.unique_id: tool for tool in _builtin_tools}
+
+    async def options_chat(request: Request) -> Response:
+        """Handle CORS preflight requests."""
+        return Response()
+
+    async def configure_frontend(request: Request) -> Response:
+        """Endpoint to configure the frontend with available models and tools."""
+        config = _ConfigureFrontend(
+            models=_models,
+            builtin_tools=[BuiltinToolInfo(id=tool.unique_id, name=tool.label) for tool in _builtin_tools],
+        )
+        return JSONResponse(config.model_dump(by_alias=True))
+
+    async def health(request: Request) -> Response:
+        """Health check endpoint."""
+        return JSONResponse({'ok': True})
+
+    async def post_chat(request: Request) -> Response:
+        """Handle chat requests via Vercel AI Adapter."""
+        adapter = await VercelAIAdapter.from_request(request, agent=agent)
+        extra_data = _ChatRequestExtra.model_validate(adapter.run_input.__pydantic_extra__)
+
+        # Validate model is in allowed list
+        if extra_data.model and _model_ids and extra_data.model not in _model_ids:
+            return JSONResponse(
+                {'error': f'Model "{extra_data.model}" is not in the allowed models list'},
+                status_code=400,
+            )
+
+        request_builtin_tools = [
+            _tools_by_id[tool_id] for tool_id in extra_data.builtin_tools if tool_id in _tools_by_id
+        ]
+        streaming_response = await VercelAIAdapter.dispatch_request(
+            request,
+            agent=agent,
+            model=extra_data.model,
```

Comment:
If a model instance was provided, we should use it instead of passing in the ID directly! We need to have a map of model IDs to the original thing that was provided (model instance or name)